test/ech_test.c

```diff
@@ -1032,12 +1131,496 @@ static int ech_test_file_read(int run)
+        && !TEST_true(SSL_set1_groups_list(serverssl, "P-384")))
+        goto end;
+    if (!TEST_true(SSL_set_tlsext_host_name(clientssl, "server.example")))
+        goto end;
+    if (!TEST_true(create_ssl_connection(serverssl, clientssl,
+                                         SSL_ERROR_NONE)))
+        goto end;
+    serverstatus = SSL_ech_get1_status(serverssl, &sinner, &souter);
+    if (verbose)
+        TEST_info("server status %d, %s, %s", serverstatus, sinner, souter);
+    if (0 && !TEST_int_eq(serverstatus, SSL_ECH_STATUS_SUCCESS))
+        goto end;
+    /* override cert verification */
+    SSL_set_verify_result(clientssl, X509_V_OK);
+    clientstatus = SSL_ech_get1_status(clientssl, &cinner, &couter);
+    if (verbose)
+        TEST_info("client status %d, %s, %s", clientstatus, cinner, couter);
+    if (0 && !TEST_int_eq(clientstatus, SSL_ECH_STATUS_SUCCESS))
+        goto end;
+    /* all good */
+    if (combo == OSSL_ECH_TEST_BASIC
+        || combo == OSSL_ECH_TEST_HRR
+        || combo == OSSL_ECH_TEST_CUSTOM) {
+        res = 1;
+        goto end;
+    }
+    /* continue for EARLY test */
+    if (combo != OSSL_ECH_TEST_EARLY)
+        goto end;
+    /* shutdown for start over */
+    sess = SSL_get1_session(clientssl);
+    OPENSSL_free(sinner);
+    OPENSSL_free(souter);
+    OPENSSL_free(cinner);
+    OPENSSL_free(couter);
+    sinner = souter = cinner = couter = NULL;
+    SSL_shutdown(clientssl);
+    SSL_shutdown(serverssl);
+    SSL_free(serverssl);
+    SSL_free(clientssl);
+    serverssl = clientssl = NULL;
+    /* second connection */
+    if (!TEST_true(create_ssl_objects(sctx, cctx, &serverssl,
+                                      &clientssl, NULL, NULL)))
+        goto end;
+    if (!TEST_true(SSL_set_tlsext_host_name(clientssl, "server.example")))
+        goto end;
+    if (!TEST_true(SSL_set_session(clientssl, sess)))
+        goto end;
+    if (!TEST_true(SSL_write_early_data(clientssl, ed, sizeof(ed), &written)))
+        goto end;
+    if (!TEST_size_t_eq(written, sizeof(ed)))
+        goto end;
+    if (!TEST_int_eq(SSL_read_early_data(serverssl, buf,
+                                         sizeof(buf), &readbytes),
+                     SSL_READ_EARLY_DATA_SUCCESS))
+        goto end;
+    if (!TEST_size_t_eq(written, readbytes))
+        goto end;
+    /*
+     * Server should be able to write data, and client should be able to
+     * read it.
+     */
+    if (!TEST_true(SSL_write_early_data(serverssl, ed, sizeof(ed), &written))
+            || !TEST_size_t_eq(written, sizeof(ed))
+            || !TEST_true(SSL_read_ex(clientssl, buf, sizeof(buf), &readbytes))
+            || !TEST_mem_eq(buf, readbytes, ed, sizeof(ed)))
+        goto end;
+    serverstatus = SSL_ech_get1_status(serverssl, &sinner, &souter);
+    if (verbose)
+        TEST_info("server status %d, %s, %s", serverstatus, sinner, souter);
+    if (0 && !TEST_int_eq(serverstatus, SSL_ECH_STATUS_SUCCESS))
```

Comment:
Add a `TODO(ECH):  test to enable once ... ` ?

Reply:
That's in a comment above already, at the top of `test_ech_roundtrip_helper`. Reasons to leave it there are a) it's temporary for maybe 2 or 3 PRs, and 2) in the next PR there are more such cases added (also temporarily). Basically all instances of `if (0 & ...`  are the same, so don't really see a need to have a `TODO(ECH)` for each.

Reply:
I prefer to have a comment close to the point of concern, that's all.

Moreover, some CI target with strict options will fail because this line will raise a warning like "Test is always false"

Reply:
CI isn't an issue here as the current ones pass and the `if (0 &` code will soon disappear. (Well, "soon" if we can get these PRs merged a bit quicker, but in any case within 2 PRs from now.) So no change is needed here.